### PR TITLE
Avatar: Don't log failure to add existing item to cache

### LINKED ISSUE #19946: Avatar: Warning incorrectly being logged
**What happened**:
Each time the avatar is fetched for my admin user a warning is logged:
`WARN[10-22|10:50:22] Error adding avatar to cache: Item 46d229b033af06a191ff2267bca9ae56 already exists`

**What you expected to happen**:
No warning logged since this is expected using the default admin account (admin@localhost) since it doesn't have a gravatar configured.

**How to reproduce it (as minimally and precisely as possible)**:
Login to Grafana using default admin account and inspect Grafana server log.

**Anything else we need to know?**:
This probably happens with all email addresses that don't have a configured gravatar.

**Environment**:
- Grafana version: master (8d6a2c2)

### PR DESCRIPTION
**Pull request by aknuds1**
**What this PR does / why we need it**:
Don't log failure to add an item to cache which is already present.

**Which issue(s) this PR fixes**:

Fixes #19946.

### CODE REVIEW

File: pkg/api/avatar/avatar.go [hunk 1]
```diff
@@ -105,7 +105,7 @@ func (this *CacheServer) Handler(ctx *macaron.Context) {
 	if avatar.notFound {
```
Comment (marefr) [inline pkg/api/avatar/avatar.go]: Feels like this is not working as expected or has stopped working sometime ago, but not sure. What is the content of `avatar` here?

**Review by marefr — CHANGES_REQUESTED** [review]
Need to do some more investigations here. Code seems to try and handle the case when avatars are missing so adding it to cache shouldn't happen in first plan (I think). Let me know if you need any help.

File: pkg/api/avatar/avatar.go (hunk 1 above)
Comment (aknuds1) [reply pkg/api/avatar/avatar.go]: @marefr What do you think isn't working here? I think at this point, the avatar can either be found or not. If it's found, it should apparently be added to the cache. Maybe we should avoid adding it to the cache though if it was originally retrieved from there?

Comment (aknuds1) [reply pkg/api/avatar/avatar.go]: I might be able to come up with a cleaner logic, where we only try to add the avatar to the cache if it wasn't already found there.

Comment (marefr) [reply pkg/api/avatar/avatar.go]: Cool :+1:

File: pkg/api/avatar/avatar.go [hunk 2]
```diff
@@ -104,8 +105,8 @@ func (this *CacheServer) Handler(ctx *macaron.Context) {
 
 	if avatar.notFound {
```
Comment (aknuds1) [inline pkg/api/avatar/avatar.go]: @marefr I reviewed the logic carefully now, and reached the conclusion that `avatar` only gets added to the cache if it was found when consulting the server (the server gets consulted for expired avatars in the cache and avatars not found in cache) and not already in cache.

If the existing logic (described above) is correct, I'm not sure there's so much we can do to improve the code here. I'd be interested in your thoughts on this.

File: pkg/api/avatar/avatar.go [hunk 3]
```diff
@@ -104,8 +105,8 @@ func (this *CacheServer) Handler(ctx *macaron.Context) {
...
 		avatar = this.notFound
-	} else {
-		if err := this.cache.Add(hash, avatar, gocache.DefaultExpiration); err != nil {
+	} else if !exists {
```
Comment (aknuds1) [inline pkg/api/avatar/avatar.go]: @marefr Could this function be called by several threads at once? In that case, using this variable is pretty pointless.

**Review by marefr — COMMENTED** [review]
Pretty sure you're right, see comment. If you have an idea of how to fixing it, please do. Otherwise we can discuss the possible solutions.

**Comment by aknuds1** [comment]
> Pretty sure you're right, see comment. If you have an idea of how to fixing it, please do. Otherwise we can discuss the possible solutions.

Thanks @marefr. I don't have any suggestions yet, but let's think through it.

**Comment by marefr** [comment]
@aknuds1 have been looking into this again and trying to understand what the code actually does. One thing that I didn't understand before is that the cache expiration time is 1 hour, but the expiration check for avatar is 10 minutes. Since a pointer to avatar is stored in cache, the `avatar.Update()` actually updates the avatar already in cache. So the `!exists` check added should be a proper solution. Could happen that avatar is removed from cache since checking if it exists and after it has been updated, but then there will be one additional update/fetch next time avatar is requested which should be fine I think.

Pushed an update and I would like to merge this for now so that we can include it in v6.5, do you agree?

**Comment by aknuds1** [comment]
@marefr You're probably right :)

**Review by marefr — APPROVED** [review]
LGTM

**Pull request merged**